ouroboros-consensus-cardano/src/tools/Cardano/Node/Protocol/Cardano.hs

```diff
@@ -140,96 +139,98 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
-          -- is in the Shelley era. That is, it is the version of protocol
-          -- /after/ Shelley, i.e. Allegra.
-          shelleyProtVer =
-            ProtVer (natVersion @3) 0,
-          shelleyMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
-        Consensus.ProtocolParamsAllegra {
-          -- This is /not/ the Allegra protocol version. It is the protocol
-          -- version that this node will declare that it understands, when it
-          -- is in the Allegra era. That is, it is the version of protocol
-          -- /after/ Allegra, i.e. Mary.
-          allegraProtVer =
-            ProtVer (natVersion @4) 0,
-          allegraMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
-        Consensus.ProtocolParamsMary {
-          -- This is /not/ the Mary protocol version. It is the protocol
-          -- version that this node will declare that it understands, when it
-          -- is in the Mary era. That is, it is the version of protocol
-          -- /after/ Mary, i.e. Alonzo.
-          maryProtVer = ProtVer (natVersion @5) 0,
-          maryMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
-        Consensus.ProtocolParamsAlonzo {
-          -- This is /not/ the Alonzo protocol version. It is the protocol
-          -- version that this node will declare that it understands, when it
-          -- is in the Alonzo era. That is, it is the version of protocol
-          -- /after/ Alonzo, i.e. Babbage.
-          alonzoProtVer = ProtVer (natVersion @6) 0,
-          alonzoMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
-        Praos.ProtocolParamsBabbage {
-          -- This is /not/ the Babbage protocol version. It is the protocol
-          -- version that this node will declare that it understands, when it
-          -- is in the Babbage era.
-          Praos.babbageProtVer = ProtVer (natVersion @7) 0,
-          Praos.babbageMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
-        Praos.ProtocolParamsConway {
-          -- This is /not/ the Conway protocol version. It is the protocol
-          -- version that this node will declare that it understands, when it
-          -- is in the Conway era.
-          Praos.conwayProtVer =
-            if npcTestEnableDevelopmentHardForkEras
-            then ProtVer (natVersion @9) 0  -- Advertise we can support Conway
-            else ProtVer (natVersion @8) 0, -- Otherwise we only advertise we know about Babbage
-          Praos.conwayMaxTxCapacityOverrides =
-            Mempool.mkOverrides Mempool.noOverridesMeasure
-        }
+            -- This is /not/ the Byron protocol version. It is the protocol
+            -- version that this node will use in blocks it creates. It is used
```

Comment:
> It is the protocol version that this node will use in blocks it creates
            
I think this sentence has cause so much confusion over the years. If I recall correctly, "use in blocks it creates" means "There's a header field that will be set to this value. Nothing else about the block depends on that field's value, not even validation. It's merely for out-of-band signaling".

I wonder if we could refine our API to make this oddness much more obvious and less of a hazard.

Reply:
Maybe that comment is more relevant to the hypothetical Issue you proposed about versions.

Reply:
Just recording what we briefly talked about in the Office Hours: The version of the *last* era will also be used to set the `MaxMajorProtVer`. The Shelley ledger specs (see Figure 74, function `chainChecks`) require that all blocks are considered invalid if the on-chain protocol version is larger than the `MaxMajorProtVer`, quoting:

> The constant `MaxMajorPV` provides a mechanism for halting outdated nodes. Once the
major component of the protocol version in the protocol parameters exceeds this value, every subsequent block is invalid.

In our code, since Praos, this is realised here:

https://github.com/input-output-hk/ouroboros-consensus/blob/421a7eaa0860bd9d07dfeca5390bf2ad4ad7dba8/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Protocol/Praos.hs#L66-L77

Somewhat confusingly, this check does *not* depend on any data from the header (`maxpv` is a constant, and `m` comes from the ledger view/state), we have https://github.com/input-output-hk/ouroboros-network/issues/4110 to improve on this.

Reply:
> The version of the _last_ era will also be used to set the `MaxMajorProtVer`.

The "last era" is the last "HFC era", right? Because ledger also has a notion of "intra-era" discrepancies. Just something to double-check: if ledger introduces another "ledger era" that corresponds to the same "HFC era" as the previous "ledger era", might that mean that `MaxMajorProtVer` should be greater than "the version of the last HFC era"?

Reply:
To be more concrete:

https://github.com/input-output-hk/cardano-node/blob/4f310a8688bb96b93e4bab56501464136491d2b0/cardano-node/src/Cardano/Node/Protocol/Cardano.hs#L195-L217

Suppose the latest thing ledger had done was create the intra-era hard fork inside of Alonzo. In that case, the chain should be on major version 6, despite the version of the last HFC era (ie Mary's in this case) being 5.

Reply:
I opened https://github.com/input-output-hk/ouroboros-consensus/issues/324 for this.